x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigInitialNode.java

```diff
@@ -0,0 +1,352 @@
+            if (false == env.settings().getByPrefix(XPackSettings.TRANSPORT_SSL_PREFIX).isEmpty() ||
+                    false == env.settings().getByPrefix(XPackSettings.HTTP_SSL_PREFIX).isEmpty()) {
+                // zero validation for the TLS settings as well, let the node bootup do its thing
+                terminal.println(Terminal.Verbosity.VERBOSE,
+                        "Skipping security auto configuration because it appears that TLS is already configured.");
+                return;
+            }
+            // Silently skipping security auto configuration because file realm is configured
+            // Auto-configuration only requires that the file realm be enabled, and, for optimal experience,
+            // also be the first in the chain.
+            if (false == env.settings().getByPrefix(RealmSettings.realmSettingPrefix(FileRealmSettings.TYPE)).isEmpty()) {
+                terminal.println(Terminal.Verbosity.VERBOSE,
+                        "Skipping security auto configuration because it appears that a file-based realm is already configured.");
+                return;
+            }
+        }
+        final Path ymlPath = env.configFile().resolve("elasticsearch.yml");
+        // The node's yml configuration is mounted read-only, which is a sign that the configuration mustn't change.
+        // Inform that auto-configuration will not run.
+        if (false == Files.isWritable(ymlPath)) {
+            terminal.println(Terminal.Verbosity.NORMAL, String.format(Locale.ROOT, "Skipping security auto configuration because " +
+                    "the configuration file [%s] is not writable", ymlPath));
+            return;
+        }
+        // Best effort determine if the node's keystore is writable.
+        // Inform that auto-configuration will not run if keystore cannot be updated.
+        final Path keystorePath = env.configFile().resolve(KeyStoreWrapper.KEYSTORE_FILENAME);
+        if (Files.exists(keystorePath) && false == Files.isWritable(keystorePath)) {
+            terminal.println(Terminal.Verbosity.NORMAL, String.format(Locale.ROOT, "Skipping security auto configuration because " +
+                    "the node keystore file [%s] is not writable", keystorePath));
+            return;
+        }
+
+        // TODO round/truncate to day, as it is easier on the tests
+        //Rounding rounding = new Rounding.Builder(Rounding.DateTimeUnit.DAY_OF_MONTH).timeZone(ZoneOffset.UTC).build();
+        //ZonedDateTime autoConfigDate = ZonedDateTime.now(ZoneOffset.UTC).truncatedTo(ChronoUnit.DAYS);
+        final ZonedDateTime autoConfigDate = ZonedDateTime.now(ZoneOffset.UTC);
+        final String instantAutoConfigName = "auto_generated_on_" + autoConfigDate.toInstant().getEpochSecond();
+        final Path instantAutoConfigDir = env.configFile().resolve(instantAutoConfigName);
+        try {
+            Files.createDirectory(instantAutoConfigDir);
```

Comment:
We should be more explicit with the permissions that we create the auto config directory with. Given that the keystores are password protected, we could go with read/write for owner and group and read for others?
```
Files.createDirectory(instantAutoConfigDir, PosixFilePermissions.asFileAttribute(
                PosixFilePermissions.fromString("drwxrwxr-x")));
```